Transac! MVP

xaun opened the pull request:
@alexnoox please review - probably should've done this in smaller pieces from the start, I apologise for that. Regarding architecture I took inspiration from the latest Todd Moto style guide & FLUX pattern. Also the gulp is very simple, it's definitely on my todo to improve this.

Thanks

FYI @BrunoChauvet

xaun committed:
Project init with basic build for Angular library

xaun committed:
Refactor project to a modular component architecture

xaun committed:
Add gulp ngAnnotate to build

xaun committed:
Add top-bar common component and style

xaun committed:
Add menus list constant for top-bar module & flexbox layout

xaun committed:
Emit menu item on click to transac component

xaun committed:
Use Transactions mock data, add Transaction cmp w/ actions

xaun committed:
Add icons & icon colors for transaction actions

xaun committed:
Adjust button spacing for bigger icon size

xaun committed:
Change transactions to white background + style cleanups

xaun committed:
Add expanded section to transaction

xaun committed:
Refactor transaction less into component css style

xaun committed:
Refactor top-bar into component css style

xaun committed:
Add globals for removing Bootstrap row & column gutters

xaun committed:
Apply .no-gutters class to transaction detailed cols

xaun committed:
Remove .btn from transcation summary-actions & add glow on-hover

xaun committed:
Add changes & app sharing section to transaction-detail

xaun committed:
Add detail section actions

xaun committed:
Update mappings on transcation actions

xaun committed:
Transaction duplicates section & adjusted mappings params behaviours

xaun committed:
Add Transaction onCommit event

xaun committed:
Add border-bottom for selected transaction

xaun committed:
Add transactions counter to top-bar

xaun committed:
Small refactors in transaction cmp & svc

xaun committed:
Add merge component rendered onTransactionMerge

xaun committed:
Add merge cmp back button

xaun committed:
Flex layout for merge component & remove transaction service

xaun committed:
Format merge transaction data & display in table layout

xaun committed:
Refactor project structure

Move the merge component under transactions as transaction-reconcile,
and rename the action of 'merging' as 'reconciling'.

This refactor unifies the reconcile view as apart of the transactions component
allowing it to use the other transactions components within itself useful because
it shares commonalities / concerns.

xaun committed:
Move details section into component + top-panel for reconcile

xaun committed:
Reconcile tx tile topbar (select tx)

xaun committed:
Update README todos

xaun committed:
Select transaction to reconcile

xaun committed:
Downgrade angular bootstrap & toastr

Resolves resolutions with mnoe-angular.

xaun committed:
Add dist files

xaun committed:
Fix undefined opts in tx svc

xaun committed:
Build dist

xaun committed:
Find selected transaction from checkbox model

xaun committed:
Add transaction-tile component for reconcile screen

xaun committed:
Improve transaction-tile styles

xaun committed:
Add review transaction-tile & next only with a selectedTx

xaun committed:
Fix transaction-tile .no-click less

xaun committed:
Improve reconcile screen transaction-tile flex display

xaun committed:
Add reconciled record preview title & subtitle

xaun committed:
Update stubbed json

xaun committed:
Add user component with user service

xaun committed:
Fetch User on transac.component init & add app loading

xaun committed:
Add getCurrentOrg method to user service

xaun committed:
Retrieve txs & matches from API

Retrieve with sso session, or configure transactions module with
basic auth creds as a developer

xaun committed:
Fix tx-reconcile flexbox grid problems for tx-tile cmps

xaun committed:
Add done button for tx-reconcile & clean up next btns

xaun committed:
Render tx-tile checkbox if isOnSelectDefined

xaun committed:
Connect merge & commit to API

xaun committed:
Add loading for transactions

xaun committed:
Build dist

xaun committed:
Add ellipsis for tx-tile topbar text subtitle

xaun committed:
Fix ng-repeat track by with duplicate tx mappings

xaun committed:
Update README

xaun committed:
Namespace everything & use BEM less convention everywhere

xaun committed:
Clean up naming & comments in transactions.service

xaun committed:
Transaction icons

xaun committed:
Tx summary subtitle, dates & changes formatting

xaun committed:
Add jsdocs to transactions.service and transaction-tile.component

xaun committed:
Fix TransacTxService promise error handlers

xaun committed:
Refactor tx cmp adding private loadTxs method

xaun committed:
Fix tx reconcile tiles to display match tx app_name

xaun committed:
Remove ngSanitize

xaun committed:
Add endless scroll pagination for transactions

xaun committed:
Add jsdoc style docs across angular files

xaun committed:
Add transaction-matches.component, formatting match titles

xaun committed:
Connect top-bar menu to transactions cmp

xaun committed:
Add transactions-history json stub

xaun committed:
Add search transactions feature

xaun committed:
Reduce search bar flex width

xaun committed:
Update txs manual-load element text "scroll for more"

xaun committed:
Quickfix search with no results not including query on retry

xaun committed:
Build dist

xaun committed:
Move top-bar to components from common

xaun committed:
Build dist

xaun committed:
Fix fail to retrieve orgUid in transactions.svc

xaun committed:
Fix loadMore txs not including search $filter

xaun committed:
Fix pagination page count not reset on menu change

xaun committed:
Build dist

xaun committed:
Fix "potential duplicates" flexbox positioning

xaun committed:
Build dist

xaun committed:
Build dist

xaun committed:
Display 'coming soon' on history txs view

xaun committed:
Use getter in user svc on fetch

xaun committed:
Remove txs on commit instantly

xaun committed:
Prevent selecting checkbox input directly on tx tile

xaun committed:
Build dist

xaun committed:
Refactor animate on commit txs code & add comments

xaun committed:
Fix empty tx getting added to txs collection on load

- Txs svc @get method rejects if response has no data.
- Add default response types for @get method

xaun committed:
Build dist

xaun committed:
Fix pagination not disabled on pagination.total undefined

xaun committed:
Add initTransactions action with TxsStore

xaun committed:
Add loading to initTransactions action

xaun committed:
Update some variable naming

xaun committed:
Add paginateTxs action to TransacTxsActions service

xaun committed:
Add reloadTxs action to TransacTxsActions service

xaun committed:
Remove loadTxs function from txs cmp

xaun committed:
Add comments & remove logs

xaun committed:
Update README

xaun committed:
Add commitTx & mergeTxs actions to TransacTxsActions

xaun committed:
Increase pagination limit to 20

xaun committed:
Merge pull request #1 from xaun/flux-pattern

Flux pattern

xaun committed:
Update menu count instantly on commit

Emit state changes to parent components on txStore subscrition notifications.

xaun committed:
Rename TransacTxsActions to TransacTxsDispatcher

xaun committed:
Update menu count on commit manually on dispatcher notify

xaun committed:
Fix cachedParams being cleared at wrong time

xaun committed:
Remove txs cmp onInit callback, call reload on txsDispatcher from transac cmp

xaun committed:
Add txs cmp txsType to txStore

xaun committed:
Improve txs manual-load element behaviour & display

xaun committed:
Disable & animate menu on txs loading

xaun committed:
Fix TxsDispatcher not properly rejecting deferred

xaun committed:
Improve TxsService formatting methods

xaun committed:
Build dist

xaun committed:
Merge screen 'grey out' unselected transactions

xaun committed:
Change merge tx tiles to fixed height

xaun committed:
Move keyframes animations to globals.less

xaun committed:
Improve merge view action buttons

xaun committed:
Add "no results" and "all results" to scroll-more section

xaun committed:
Refactor pagination system by using api next url

xaun committed:
Fix menu txs count not updated on search

xaun committed:
Fix txs service formatAttribute bug

xaun committed:
Add transaction title & subtitle flexbox ellipsis

xaun committed:
Add padding to transac-tx-maches .match-caption

xaun committed:
Build dist

xaun committed:
Add Transactions filters selector

xaun committed:
Disable search-bar on txs loading

xaun committed:
Add toastr alerts via TransacAlertsService

xaun committed:
Fix minor bug in top-bar cmp $onChanges

xaun committed:
Refactor top-bar to use flux pattern

xaun committed:
Merge pull request #2 from xaun/flux-topbar

Refactor top-bar to use flux pattern

xaun committed:
Build dist

xaun committed:
Fix $filter query format

xaun committed:
Build dist

xaun committed:
Fix transaction summary <a> nesting ng-click issue

xaun committed:
Fix transaction changes height

xaun committed:
Add strong box shadow on selected tx

xaun committed:
Improve secondary action buttons alike mockups

xaun committed:
Add tooltip to summary actions

xaun committed:
Scroll to transaction on merge screen back

xaun committed:
Add scroll top of page on merge screen

xaun committed:
Add tx-icon global classes

xaun committed:
Improve transaction layout & action buttons

xaun committed:
Add flat class for tx-icon

xaun committed:
Add refresh txs / hist top-bar button

xaun committed:
Scroll to tx on select, deselect all on 'esc' keydown

xaun committed:
Fix subtitle displaying 'to' when there are no toApps

xaun committed:
Add overflow auto to tx-tile tx-changes

xaun committed:
Only including pending mappings in share with selection

xaun committed:
Build dist

xaun committed:
Change tx actions to tx-fab icons

xaun committed:
Enable historical txs

xaun committed:
Fix filters not applied on refresh

xaun committed:
Build dist

xaun committed:
Fetch historical count for top-bar on app load

xaun committed:
Update readme

xaun committed:
Add include / exclude duplicates filter

xaun committed:
Adjust codebase to use new matching_records response

xaun committed:
Build dist

xaun committed:
Fix duplicates filter bugs

xaun committed:
Add reset states on txs & topbar cmp scope $onDestroy

States need to be managed $onDestroy to prevent duplicate objects / conflicting data

xaun committed:
Remove json stub files

xaun committed:
Add disable pointer for topbar refresh btn on loading

xaun committed:
Add recursive load lock for reloadTxs

Parent application may call reloadTxs while currently reloading, this prevents data state
duplicates / issues.

xaun committed:
Build dist

xaun committed:
Fix load pagination totals before fetched user

xaun committed:
Build dist

xaun committed:
Add interval to gulp watch

xaun committed:
Clean up naming across the app

xaun committed:
Fix txs actions overflowing on sm screens

xaun committed:
Add desc for top-bar dispatcher file

xaun committed:
Remove BEM syntax across the components

xaun committed:
Build dist

xaun committed:
Safari display fixes

xaun committed:
Clean up toastrConfig

xaun committed:
Build dist

alexnoox commented on src/transac/common/common.module.coffee:
```diff
@@ -0,0 +1,7 @@
+###
+angular.module('transac.common',
+  [
+    #deps
```
This module doesn't seem to be used. Can be removed?

xaun replied:
It is a placeholder following Todd Motto's file structure, I sometimes feel leaving this means if other developers want to add something, it may act as a hint.. But I'm not sure. Should I remove it?

alexnoox commented on src/transac/components/alerts/alerts.module.coffee:
```diff
@@ -0,0 +1,23 @@
+    info:
+      timeOut: 800
+    error:
+      timeOut: 1500
```
If these timeouts are in milliseconds, they must appear really briefly.

xaun replied:
Yes that's by design, notifications are sent when you successfully allow / deny transactions, and you can do this very quickly, so long durations would cause the notifications to build up very fast. The error timeOut is longer - and could be longer if preferred.

alexnoox commented on src/transac/components/alerts/alerts.service.coffee:
```diff
@@ -0,0 +1,8 @@
+angular.module('transac.alerts').service('TransacAlertsService', (toastr, TOASTR_OPTS)->
```
I think we should more talk about notifications than alerts.

xaun replied:
Sure, when you say "more talk" does that mean you think I should rename it to notifications ?

alexnoox replied:
Yes you should rename it to notification

alexnoox commented on src/transac/components/components.module.coffee:
```diff
@@ -0,0 +1,10 @@
+###
+#   @desc The Components module is the container reference for all reusable modules. Modules added as depenedencies here should be decoupled.
+###
+angular.module('transac.components',
```
I think you should remove this module wrapper, and directly add the 4 modules in `maestrano.transac`. It is just adding a useless layer.

xaun replied:
I debated with myself on this one a fair bit. As I have coded Transac! in a way which each module could be removed from the project and run given the parent component's required bindings are provided.

If I put all the modules under `transac`, then it doesn't really indicate what is stand alone. But this is also mainly following todd motto's example apps and the way he seperates the angular modules. What do you think?

alexnoox replied:
I think that on the paper it is a good idea, in the reality for 4 modules it is an overkill. You should simplify this.

If Transac! codebase keeps growing then we will apply this pattern. But if it is not needed it seems to me it is adding useless complexity.

alexnoox commented on src/transac/components/top-bar/filters-selector/filters-selector.component.coffee:
```diff
@@ -0,0 +1,36 @@
+    ctrl = this;
+
+    ctrl.$onInit = ->
+      ctrl.wasSelectionMade = false
```
Should be renamed to `ctrl.isSelectionMade`.

alexnoox commented on src/transac/components/top-bar/filters-selector/filters-selector.component.coffee:
```diff
@@ -0,0 +1,36 @@
+      ctrl.onSelect(EventEmitter(selectedFilter: filter))
+
+    return
+
```
Empty line to remove

alexnoox commented on src/transac/components/top-bar/filters-selector/filters-selector.html:
```diff
@@ -0,0 +1,12 @@
+<div class="filters-selector">
+  <div class="btn-group" uib-dropdown on-toggle="$ctrl.toggleSelector(open)" auto-close="outsideClick">
+    <button id="dropdown-toggle-btn" type="button" class="btn btn-primary" ng-class="{ 'disabled': $ctrl.isDisabled }" ng-click="$ctrl.toggleSelector()" uib-dropdown-toggle>
```
You can use `ng-disabled` on a button

xaun replied:
Not sure if it was working because of `uib-dropdown-toggle` - I will take a look, thanks

alexnoox commented on src/transac/components/top-bar/search-bar/search-bar.component.coffee:
```diff
@@ -0,0 +1,42 @@
+
+    ctrl.submitOnKeypress = ($event={}, force=false)->
+      return if ctrl.isDisabled
+      return unless $event.keyCode == 13 || force
```
Add a constant `KEYCODE_ENTER = 13` to avoid a magic number.

alexnoox commented on src/transac/components/top-bar/top-bar.module.coffee:
```diff
@@ -0,0 +1,32 @@
+    {label: 'Organizations', type: '$filter', attr: 'entity_type', cmd: 'match', value: "/Organization/", selected: false}
+  ])
+  # EventEmitter wrapper for emitting events through component '&' callbacks.
+  .value('EventEmitter', (payload)-> { $event: payload })
```
This object is used across the whole project. It should be removed from the `top-bar` module to a more common module.

alexnoox commented on src/transac/components/top-bar/top-bar.store.coffee:
```diff
@@ -0,0 +1,99 @@
+        # Only one $orderby filter can be selected, and atleast one is always selected
+        if payload.type == '$orderby'
+          filters = _.filter(state.filtersMenu, 'type', payload.type)
+          _.each(filters, (f)-> f.selected = false & undefined)
```
Why do you assign the value `false & undefined`? It returns `0`. Don't you want `f.selected` to be set at `false`?

xaun replied:
Yes, this looks like a mistake. The `_.each` method breaks when false is returned, which annoying with coffeescript one liners.

alexnoox replied:
`_.each(filters, (f) -> f.selected = false; true)` should work

alexnoox commented on src/transac/components/transactions/transaction-changes/transaction-changes.component.coffee:
```diff
@@ -0,0 +1,18 @@
+    onSelect: '&?'
+  }
+  templateUrl: 'components/transactions/transaction-changes'
+  controller: (EventEmitter)->
```
This controller does nothing you can remove it.

alexnoox commented on src/transac/components/transactions/transaction-matches/transaction-matches.component.coffee:
```diff
@@ -0,0 +1,24 @@
+    ctrl.subtitle = (match)->
+      # appName = _.get(match, 'app_name')
+      date = TransacTxsService.formatDisplayDate(_.get(match, 'created_at'))
+      # if appName then "#{date}, from #{appName}" else date
```
Avoid commented code

alexnoox commented on src/transac/components/transactions/transactions.module.coffee:
```diff
@@ -0,0 +1,25 @@
+  .value('THROTTLE_MILLISECONDS', 1000)
+  # Add Connec API! basic auth keys - DO NOT COMMIT.
+  # Note: this is temporary while the app is in early stages of developement.
+  .constant('DEV_AUTH',
```
Can you add a mechanism to configure Connec! credentials?

xaun replied:
Any suggestions?

alexnoox replied:
Transac! could provide a `TransacAuthService`, configured by the host library `mno-enterprise-angular`.

alexnoox commented on src/transac/components/transactions/transactions.module.coffee:
```diff
@@ -0,0 +1,25 @@
+    'ngAnimate'
+  ])
+  # EventEmitter wrapper for emitting events through component '&' callbacks.
+  .value('EventEmitter', (payload)-> { $event: payload })
```
Cf. previous comment.

alexnoox commented on src/transac/components/transactions/transactions.service.coffee:
```diff
@@ -0,0 +1,255 @@
+
+  _self = @
+
+  _self.HTTP_CONFIG = {}
```
By convention, the constants are in uppercase. This is not a constant, it should be camel cased.

alexnoox commented on src/transac/components/transactions/transactions.service.coffee:
```diff
@@ -0,0 +1,255 @@
+    )
+
+  ###
+  #   @desc [DEPRECIATED] Find matching transacations rated with a score representing
```
Typo: `DEPRECATED `

alexnoox replied:
`depreciated (verb): diminish in value over a period of time.`

alexnoox commented on src/transac/components/transactions/transactions.service.coffee:
```diff
@@ -0,0 +1,255 @@
+  #   @param {Object} [x] Object to flatten.
+  #   @returns {Object} Flattened object.
+  ###
+  @flattenObject = (x, result = {}, prefix = null)->
```
Would simpler to use lodash https://lodash.com/docs/#flattenDeep?

alexnoox commented on bower.json:
```diff
@@ -0,0 +1,39 @@
+    "bootstrap": "^3.3.6",
+    "lodash": "~3.10.1",
+    "font-awesome": "~4.7.0",
+    "moment": "~2.11.2",
```
You should use https://github.com/urish/angular-moment, it is an angular wrapper.

alexnoox commented on src/transac/components/transactions/transactions.dispatcher.coffee:
```diff
@@ -0,0 +1,132 @@
+      $timeout(->
+        $log.warn("Reload Transactions locked: retrying in 5sec")
+        _self.reloadTxs(options)
+      ,500)
```
This is half a second, not 5 seconds.

alexnoox commented on src/transac/components/transactions/transactions.dispatcher.coffee:
```diff
@@ -0,0 +1,132 @@
+  #   @params See service method comments
+  #   @returns {Promise<Object>} whether the commit was successful or not
+  ###
+  @commitTx = (url, resource, mappings)->
```
Why don't you simply pass the transaction object?

alexnoox commented on src/transac/components/transactions/transactions.dispatcher.coffee:
```diff
@@ -0,0 +1,132 @@
+  @commitTx = (url, resource, mappings)->
+    deferred = $q.defer()
+    TransacTxsStore.dispatch('minusPgnTotal', 1)
+    $timeout((-> deferred.notify(true)), 0)
```
Delay is optional.

alexnoox reviewed (commented):
Hi @xaun,

First of all, great job! The FLUX approach is good and Todd Motto is always a great source of inspiration. Using a store to manage transactions especially make sense.

This is a first quick and dirty review. The code base is starting to be consequent and I will need to take more time to review everything.

Some general points:
- i18n to add
- Remove commented code

Happy to discuss. Also if you want me to focus on some critical parts, tell me what to look at.

alexnoox commented:
Also, could you send me some articles or anything you inspired yourself from for your FLUX implementation?

xaun commented:
@alexnoox thanks for your reviews, I have added replies to some comments, the others I will fix up when I can. 

Regarding focusing on a certain section of the code, I think definitely the dispatchers & stores. 
I would like to take a look at what libraries I can use to improve the FLUX pattern implementation. As I have implemented my own thing... My reason for this being I was having problems getting the angular 1.5 ng-redux library to work the way I wanted too, so ended up doing something myself. 

Unfortunately I don't have the articles saved, but general googling of angular 1 flux architecture, and looking at what Redux.js aims to provide.

alexnoox commented:
@x4d3 Ok I will have a look at the dispatchers & stores. I will also check some FLUX libraries for AngularJS.
